src/readers/CMakeLists.txt

```diff
@@ -0,0 +1,251 @@
+# This macro generates the reader header from the F3DReaderBoilerPlate.h.in template file
+# It also adds the code in the F3DReaderInstanciator.h.in file
+macro(declare_f3d_reader)
+  cmake_parse_arguments(F3D_READER "" "NAME;VTK_IMPORTER;VTK_READER;DESCRIPTION" "EXTENSIONS;MIMETYPES;READER_CUSTOM_CODE;IMPORTER_CUSTOM_CODE" ${ARGN})
+
+  list(APPEND F3D_READER_FILES "F3D${F3D_READER_NAME}.h")
+  set(F3D_READER_INCLUDES_CODE
+    "${F3D_READER_INCLUDES_CODE}#include \"F3D${F3D_READER_NAME}.h\"\n")
+  set(F3D_READER_INSTANCES_CODE
+    "${F3D_READER_INSTANCES_CODE}  F3D${F3D_READER_NAME} InstanceOfF3D${F3D_READER_NAME};\n")
+  set(F3D_READER_REGISTER_CODE
+    "${F3D_READER_REGISTER_CODE}    F3DReaderFactory::Register(&InstanceOfF3D${F3D_READER_NAME});\n")
+  set(F3D_READER_UNREGISTER_CODE
+    "${F3D_READER_UNREGISTER_CODE}    F3DReaderFactory::Unregister(&InstanceOfF3D${F3D_READER_NAME});\n")
+
+  list(APPEND F3D_SUPPORTED_MIME_TYPES "${F3D_READER_MIMETYPES}")
+  list(APPEND F3D_SUPPORTED_EXTENSIONS "${F3D_READER_EXTENSIONS}")
+
+  list(TRANSFORM F3D_READER_EXTENSIONS PREPEND "\".")
+  list(TRANSFORM F3D_READER_EXTENSIONS APPEND "\"")
+  list(JOIN F3D_READER_EXTENSIONS ", " F3D_READER_EXTENSIONS)
+
+  list(TRANSFORM F3D_READER_MIMETYPES PREPEND "\"")
+  list(TRANSFORM F3D_READER_MIMETYPES APPEND "\"")
+  list(JOIN F3D_READER_MIMETYPES ", " F3D_READER_MIMETYPES)
+
+  if(F3D_READER_VTK_IMPORTER)
+    set(F3D_READER_HAS_SCENE_READER 1)
+  else()
+    set(F3D_READER_HAS_SCENE_READER 0)
+  endif()
+
+  if(F3D_READER_VTK_READER)
+    set(F3D_READER_HAS_GEOMETRY_READER 1)
+  else()
+    set(F3D_READER_HAS_GEOMETRY_READER 0)
+  endif()
+
+  configure_file("${CMAKE_CURRENT_SOURCE_DIR}/F3DReaderBoilerPlate.h.in"
+    "${CMAKE_CURRENT_BINARY_DIR}/F3D${F3D_READER_NAME}.h")
+endmacro()
+
+declare_f3d_reader(
+  NAME 3DSReader
+  EXTENSIONS 3ds
+  MIMETYPES application/vnd.3ds
+  VTK_IMPORTER vtk3DSImporter
+  DESCRIPTION "3DS files reader"
+)
+
+declare_f3d_reader(
+  NAME CityGMLReader
+  EXTENSIONS gml
+  MIMETYPES application/gml+xml
+  VTK_READER vtkCityGMLReader
+  DESCRIPTION "CityGML files reader"
+)
+
+declare_f3d_reader(
+  NAME DICOMReader
+  EXTENSIONS dcm
+  MIMETYPES application/dicom
+  VTK_READER vtkDICOMImageReader
+  DESCRIPTION "DICOM images files reader"
+)
+
+declare_f3d_reader(
+  NAME GLTFReader
+  EXTENSIONS gltf glb
+  MIMETYPES model/gltf-binary model/gltf+json
+  VTK_READER vtkGLTFReader
+  VTK_IMPORTER vtkGLTFImporter
+  DESCRIPTION "glTF files reader"
+  READER_CUSTOM_CODE
+    "// Enable all animations in the GLTFReader
```

Comment:
Wow! I like code generation but putting C++ production code into a CMake file looks like a very uncommon (unpleasant) idea to me. It makes things looks like simpler at some levels, but can make them difficult to handle at others. I especially think to the prototyping/development steps, and future improvements/new features we could have to add to the readers.

Reply:
It's still a draft. I'm thinking of potential solution as I also think it's the weak point of this change.

Reply:
OK. I guess the header and some simple methods could be generated but the core reading/importing code could still stay in dedicated c++ files?

Reply:
custom code could be in dedicated pseudo cxx.in file to be read by cmake using FILE(READ) ?

Reply:
alternatively we handle dedicated logic in vtk* class dedicated to f3d.

Reply:
I think cxx.in file would be boring. I guess there would not be any need to read the code file in cmake, it could just be cxx files compiled and linked with the rest of the library, beside the autogenerated code.

Reply:
I have created `.inl` files, which is quite usual for files that are intented to be included in header files (not part of a compilation unit).